### PR TITLE
fwupd: add package

### PR DESCRIPTION
**Pull request by lvoegl**
Maintainer: me
Compile tested: x86_64, lantiq_xrx200
Run tested: lantiq_xrx200, 23.05

Description:
Adds [fwupd](https://github.com/fwupd/fwupd) and its four required dependecies [libjcat](https://github.com/hughsie/libjcat), [libxmlb](https://github.com/hughsie/libxmlb), [libgusb](https://github.com/hughsie/libgusb) and [libgudev](https://gitlab.gnome.org/GNOME/libgudev).

fwupd allows firmware upgrades for various devices, including many modems via the modem-manager and fastboot options. The package includes a daemon `/usr/libexec/fwupd/fwupd` started by `/etc/init.d/fwupd` and controlled by `/usr/bin/fwupdmgr`, as well as  `/usr/bin/fwupdtool` that can be used without the daemon.

### CODE REVIEW

File: libs/libgusb/Makefile [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=libgusb
+PKG_VERSION:=0.4.9
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=https://github.com/hughsie/libgusb/releases/download/$(PKG_VERSION)
+PKG_SOURCE_VERSION:=$(PKG_VERSION)
```
Comment (BKPepe) [inline libs/libgusb/Makefile:16]: This is not used.

File: libs/libjcat/Makefile [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=libjcat
+PKG_VERSION:=0.2.1
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=https://github.com/hughsie/libjcat/releases/download/$(PKG_VERSION)
+PKG_SOURCE_VERSION:=$(PKG_VERSION)
```
Comment (BKPepe) [inline libs/libjcat/Makefile:16]: Can be dropped

File: libs/libjcat/Config.in [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	help
+	  Compile libjcat with GPG support
+
+config LIBJCAT_PKCS7
+	bool "PKCS #7"
+	default y
+	help
+	  Compile libjcat with PKCS #7 support
+
+config LIBJCAT_GPG
```
Comment (BKPepe) [inline libs/libjcat/Config.in:13]: This does not look correct. :) Should be ED25519

File: libs/libgudev/Makefile [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/nls.mk
+include $(INCLUDE_DIR)/meson.mk
+
+define Package/libgudev
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=libgudev
+  URL:=https://wiki.gnome.org/Projects/libgudev
```
Comment (BKPepe) [inline libs/libgudev/Makefile:33]: I checked and that website is gonna be retired soon. What I have checked is that the official website is their GitLab.
```suggestion
  URL:=https://gitlab.gnome.org/GNOME/libgudev
```

File: libs/libgusb/Makefile [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+define Package/libgusb
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=libgusb
+  DEPENDS:= \
+	$(INTL_DEPENDS) \
+	+glib2 \
+	+libusb-1.0 \
+	+json-glib
```
Comment (BKPepe) [inline libs/libgusb/Makefile:37]: Would be great to have here URL as you have it for other packages.

File: libs/libgusb/Makefile [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  GUsb is a GObject wrapper for libusb1 that makes it easy to do asynchronous
+  control, bulk and interrupt transfers with proper cancellation and
+  integration into a mainloop.
+endef
+
+MESON_ARGS += \
+	-Db_lto=true \
+	-Dtests=false \
+	-Dvapi=false \
+	-Dusb_ids=/usr/share/hwdata/usb.ids \
```
Comment (BKPepe) [inline libs/libgusb/Makefile:50]: Shouldnt this depend somehow on ``hwdata`` package, I mean ``usbids``?

File: libs/libgusb/Makefile [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libgusb.so* $(1)/usr/lib
+
+	$(INSTALL_DIR) $(1)/usr/lib/pkgconfig
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/gusb.pc $(1)/usr/lib/pkgconfig
+endef
+
+define Package/libgusb/install
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(CP) $(PKG_INSTALL_DIR)/usr/bin/gusbcmd $(1)/usr/bin
```
Comment (BKPepe) [inline libs/libgusb/Makefile:71]:
```suggestion
	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/gusbcmd $(1)/usr/bin
```

File: libs/libjcat/Makefile [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/nls.mk
+include $(INCLUDE_DIR)/meson.mk
+
+define Package/libjcat
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=libjcat
+  DEPENDS:= \
```
Comment (BKPepe) [inline libs/libjcat/Makefile:33]: URL for this package, please. :)

File: libs/libxmlb/Config.in [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+	bool "LZMA"
+	default n
+	help
+	  Compile libxmlb with LZMA support
+
+config LIBXMLB_ZSTD
+	bool "zstd"
+	default n
+	help
+	  Compile libxmlb with Zstandard support
```
Comment (BKPepe) [inline libs/libxmlb/Config.in:14]:
```suggestion
	  Compile libxmlb with zstd support
```

File: libs/libxmlb/Makefile [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+define Package/libxmlb
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=libxmlb
+  DEPENDS:= \
+	$(INTL_DEPENDS) \
+	+glib2 \
+	+LIBXMLB_LZMA:liblzma \
+	+LIBXMLB_ZSTD:libzstd
+endef
```
Comment (BKPepe) [inline libs/libxmlb/Makefile:38]: URL here as well, please.

File: libs/libgusb/Makefile (hunk 6 above)
Comment (lvoegl) [reply libs/libgusb/Makefile:50]: Correct. `usbids` installs `/usr/share/hwdata/usb.ids`.

**Comment by lvoegl** [comment]
Added missing `PKG_CPE_ID` for fwupd.

File: libs/libxmlb/Makefile [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=https://github.com/hughsie/libxmlb/releases/download/$(PKG_VERSION)
+PKG_HASH:=0a3ec258b12dbf10e5fe840b2421c84137eb7cc1b09c3de6210f3f7d51733733
+
+PKG_MAINTAINER:=Lukas Voegl <lvoegl@tdt.de>
+PKG_LICENSE:=LGPL-2.1-or-later
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_DEPENDS:=glib2/host
+PKG_BUILD_PARALLEL:=1
```
Comment (neheb) [inline libs/libxmlb/Makefile:23]: This is not needed. It's defined in meson.mk.

File: libs/libxmlb/Makefile [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+include $(INCLUDE_DIR)/nls.mk
+include $(INCLUDE_DIR)/meson.mk
+
+define Package/libxmlb
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=libxmlb
+  URL:=https://github.com/hughsie/libxmlb
+  DEPENDS:= \
+	$(INTL_DEPENDS) \
```
Comment (neheb) [inline libs/libxmlb/Makefile:35]: INTL_DEPENDS comes with glib2.

**Comment by neheb** [comment]
Wonder if this even works given https://github.com/openwrt/packages/issues/18148

**Comment by lvoegl** [comment]
I've removed the fastboot option which removes the requirement of libgusb and therefore usbids. It can be added again with the next major version as [libgusb was dropped](https://github.com/fwupd/fwupd/commit/2c6c4dc9a51cda4baaebe1f9dac5f84811f5fd47).

Also I've decided to split the package into `fwupd-libs` `fwupd-tools` and `fwupd-daemon` so the daemon won't be automatically installed and started if not wanted.

**Comment by lvoegl** [comment]
@neheb The package is now successfully building with a working `test.sh`.

**Pull request merged**